include/glow/Runtime/RuntimeTypes.h

```diff
@@ -143,6 +144,55 @@ struct DeviceConfig {
       : backendKind(kind), name(name), parameters(parameters) {}
 
   bool hasName() const { return name != ""; }
+
+  void setDeviceMemory(uint64_t memSize) { deviceMemory = memSize; }
+
+  uint64_t getDeviceMemory() const { return deviceMemory; }
+
+  /// Handle memory setting priorites for all backends.
+  uint64_t getDeviceMemory(uint64_t defaultMemory) const {
```

Comment:
I think cleaner implementation would be:

``` cpp
uint64_t getDeviceMemory(uint64_t defaultMemory) const {
  return deviceMemory == 0 ? defaultMemory : deviceMemory;
}

```

Reply:
> I think cleaner implementation would be:
> 
> ```c++
> uint64_t getDeviceMemory(uint64_t defaultMemory) const {
>   return deviceMemory == 0 ? defaultMemory : deviceMemory;
> }
> ```

OK. I've changed to this cleaner implementation.

Reply:
Why do we need to have a switch on backendkind? Can caller provide the default value and we just return it in case deviceMemory  is not set (deviceMemory == 0).

Reply:
> Why do we need to have a switch on backendkind (this is not scalable, required updates every time we create a new backend)? Can caller provide the default value and we just return it in case deviceMemory is not set (deviceMemory == 0).

For CPU and interpreter backends, the default memory is 2GB. Basically it the following conditions.
1. we have both commandLine option(e.g. 1GB) and deviceConfigs(e.g.3GB), we should prefer commandline setting(1GB).
2. we have neither commandline nor deviceconfig, we need to stay with the default 2GB.
3. we have only commandline options(e.g. 1GB), we go with commandline option(1GB).
4. we only have deviceconfig(e.g. 3GB), we go with device config(3GB).

The habana has similar logic but different default memory size.

The OpenCL has a little bit different logic. The only difference is that the default memory size
might be changed by the device_mem retrieved from clGetDeviceInfo. If the deviceconfig is smaller than this value, we should go with deviceconfig. If the deviceConfig is larger, we should go with the clGetDeviceInfo because we do not have that much memory on device.

So that are a little bit different logics for different backends. So currently, I think we may need to implement them either in the devicemanager like previous commit or inside the deviceConfig. 
Let me know if you some other suggestions :)

Reply:
Discussed this offline.

We need to move device memory entirely to the config and make all device managers accept non optional config. This is bigger change and could be done at later point.
To unblock this PR let's proceed with the original version.